crates/jstzd/src/runnable_image/mod.rs

```diff
@@ -0,0 +1,249 @@
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    net::IpAddr,
+    sync::Arc,
+};
+
+use bollard::{
+    container::{Config, CreateContainerOptions},
+    secret::{HostConfig, PortBinding, PortMap},
+    Docker,
+};
+
+use crate::{container::ContainerId, image::Image};
+
+pub mod builder;
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum AccessMode {
+    ReadOnly,
+    ReadWrite,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum MountType {
+    Bind,
+    Volume,
+    Tmpfs,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Mount {
+    pub access_mode: AccessMode,
+    pub mount_type: MountType,
+    pub source: Option<String>,
+    pub target: Option<String>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Host {
+    Addr(IpAddr),
+    HostGateway,
+}
+
+impl ToString for Host {
+    fn to_string(&self) -> String {
+        match self {
+            Host::Addr(ip) => ip.to_string(),
+            Host::HostGateway => "host-gateway".to_string(),
+        }
+    }
+}
+
+type HostPort = u16;
+type ContainerPort = u16;
+pub struct RunnableImage<I: Image> {
+    pub image: I,
+    pub container_name: String,
+    pub overridden_cmd: Option<Vec<String>>,
+    pub network: Option<String>, //todo
+    pub env_vars: BTreeMap<String, String>,
+    pub hosts: BTreeMap<String, Host>,
+    pub mounts: Vec<Mount>, //todo
+    pub ports: BTreeMap<HostPort, ContainerPort>,
+}
```

Comment:
Similarly, the builder and runnable image are the same type. Recommend using a single type (and file) for these